example_allocators/compartment_alloc/main.c

```diff
@@ -0,0 +1,29 @@
+#include "compartment_alloc.c"
+#include "../../include/common.h"
+
+int main() {
+  // Create capabilities which the heap compartments will construct capabilities against
+  // (as if they were the DDC when the allocation is made).
+  // One will be read-only, the other write-only, so we can differentiate when printing.
+  void *__capability dc1 = cheri_perms_and(cheri_ddc_get(), CHERI_PERM_STORE);
+  void *__capability dc2 = cheri_perms_and(cheri_ddc_get(), CHERI_PERM_LOAD);
+
+  // Create two compartments, and receive the capability that gives us authority to use them
+  void *__capability compartment1 = init_compartment(4096, dc1);
+  void *__capability compartment2 = init_compartment(4096, dc2);
+
+  // Allocate memory in each compartment
+  void *__capability c1_allocated_memory = malloc_compartment(64, compartment1);
+  void *__capability c2_allocated_memory = malloc_compartment(256, compartment2);
+
+  // Print an explainer of each allocation (and the capability itself)
+  printf("\n\n\tThis capability is heap-allocated within our first compartment, which is write-only.\n\tWe allocated 64 bytes.\n");
+  pp_cap(c1_allocated_memory);
+
+  printf("\n\n\tThis capability is heap-allocated within our second compartment, which is read-only.\n\tWe allocated 256 bytes.\n");
+  pp_cap(c2_allocated_memory);
+
+  printf("\nCompleted successfully.\n");
+
+  return 0;
```

Comment:
I believe there is leaked memory at this point. Perhaps a `clean_compartment` function or so to `free` all the `mmap`s in `init_compartment`.

Reply:
Sure! I can add that. It'll need to do pointer revocation on the capabilities handed out to the memory being cleared, right? I haven't seen examples of that yet, if you have a code snippet handy I'd be curious to see (no worries if not, I'll google and implement when I get a sec).

(FWIW, this is following the lead of the bump allocator, which just has two functions analogous to the two this example contains — maybe I'll extend that too if I get a slice of time for it.)

Reply:
For an example we can just put a comment along the lines of "// In the interests of simplicity we allow the XYZ to leak." if it's not worth writing a clear-up routine. [I don't feel strongly either way.]

Reply:
>  It'll need to do pointer revocation on the capabilities handed out to the memory being cleared, right?

I'm not sure where pointer revocation comes from. I just meant to clean up the `mmap`s in `init_compartment` as it's considered a memory leak to not do so, to my knowledge (could verify with `asan` or so).

Reply:
Compartments are now freed, fixed in [bd783eb](https://github.com/capablevms/cheri-examples/pull/88/commits/bd783eb6394493b22697536ff14049ca9b5a2e2a) — how does that look?